File: ios/ApiClient.swift
import Foundation

enum ApiClient {
    struct HttpResult {
        let code: Int
        let body: String
    }

    enum ApiError: Error {
        case invalidUrl
    }

    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 3
        config.timeoutIntervalForResource = 3
        return URLSession(configuration: config)
    }()

    static func buildUrl(baseUrl: String, path: String) -> String {
        let trimmed = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
        return trimmed + path
    }

    static func executeRequest(
        method: String,
        url: String,
        jsonBody: [String: Any]? = nil,
        token: String? = nil
    ) async throws -> HttpResult {
        guard let endpoint = URL(string: url) else {
            throw ApiError.invalidUrl
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let jsonBody {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: jsonBody, options: [])
        }

        let (data, response) = try await session.data(for: request)
        let code = (response as? HTTPURLResponse)?.statusCode ?? -1
        let body = String(data: data, encoding: .utf8) ?? ""
        return HttpResult(code: code, body: body)
    }

    static func extractErrorMessage(_ body: String) -> String {
        let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return "No response body"
        }

        guard let data = body.data(using: .utf8) else {
            return body
        }

        do {
            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                let error = (json["error"] as? String ?? "")
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                return error.isEmpty ? body : error
            }
        } catch {
            return body
        }

        return body
    }
}

File: android/app/src/main/java/com/chogm/discordapp/ApiClient.kt
package com.chogm.discordapp

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object ApiClient {
    data class HttpResult(val code: Int, val body: String)

    fun buildUrl(baseUrl: String, path: String): String {
        return baseUrl.trimEnd('/') + path
    }

    fun buildWebSocketUrl(baseUrl: String, path: String): String? {
        val trimmed = baseUrl.trim().trimEnd { it == '/' }
        return when {
            trimmed.startsWith("https://") -> {
                "wss://" + trimmed.removePrefix("https://") + path
            }
            trimmed.startsWith("http://") -> {
                "ws://" + trimmed.removePrefix("http://") + path
            }
            else -> null
        }
    }

    fun executeRequest(
        method: String,
        url: String,
        jsonBody: JSONObject? = null,
        token: String? = null
    ): HttpResult {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = 5000
        connection.readTimeout = 5000
        connection.requestMethod = method
        connection.setRequestProperty("Accept", "application/json")
        if (!token.isNullOrBlank()) {
            connection.setRequestProperty("Authorization", "Bearer $token")
        }
        if (jsonBody != null) {
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.bufferedWriter().use { it.write(jsonBody.toString()) }
        }

        val code = connection.responseCode
        val body = readResponseBody(connection, code)
        connection.disconnect()
        return HttpResult(code, body)
    }

    fun extractErrorMessage(body: String): String {
        if (body.isBlank()) {
            return "No response body"
        }
        return try {
            val error = JSONObject(body).optString("error").trim()
            if (error.isBlank()) body else error
        } catch (_: Exception) {
            body
        }
    }

    private fun readResponseBody(connection: HttpURLConnection, code: Int): String {
        val stream = if (code in 200..299) connection.inputStream else connection.errorStream
        return stream?.bufferedReader()?.use { it.readText() } ?: ""
    }
}
